plugins/api/src/replica_close.cpp

```diff
@@ -0,0 +1,378 @@
+        std::strncpy(stat_input.addr.hostAddr, location.c_str(), NAME_LEN);
+
+        rodsStat_t* stat_output{};
+        if (const auto ec = rsFileStat(&_comm, &stat_input, &stat_output); ec != 0) {
+            log::api::error("Could not stat file [error_code={}, logical_path={}, physical_path={}]",
+                            ec, logical_path, physical_path);
+            return ec;
+        }
+
+        const auto size = stat_output->st_size;
+        std::free(stat_output);
+
+        return size;
+    }
+
+    auto update_replica_size_and_status(rsComm_t& _comm,
+                                        const std::string& _logical_path,
+                                        const std::string& _resource_hierarchy,
+                                        std::int64_t _new_size) -> int
+    {
+        dataObjInfo_t info{};
+        std::strncpy(info.objPath, _logical_path.data(), MAX_NAME_LEN);
+        std::strncpy(info.rescHier, _resource_hierarchy.data(), MAX_NAME_LEN);
+
+        const auto new_size_string = std::to_string(_new_size);
+        keyValPair_t reg_params{};
+        addKeyVal(&reg_params, DATA_SIZE_KW, new_size_string.data());
+        addKeyVal(&reg_params, REPL_STATUS_KW, "1");
+
+        modDataObjMeta_t input{};
+        input.dataObjInfo = &info;
+        input.regParam = &reg_params;
+
+        return rsModDataObjMeta(&_comm, &input);
+    }
+
+    auto rs_replica_close(rsComm_t* _comm, bytesBuf_t* _input) -> int
+    {
+        if (const auto [valid, msg] = is_input_valid(_input); !valid) {
+            log::api::error(msg);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        // TODO The following should happen:
+        // 1. Redirect to the correct server.
+        // 2. Close the physical file.
+        // 3. Clean up (free L1 descriptors)!
+        // 4. Stat the physical file to get the actual file size.
+        // 5. Update the data object's catalog entry with the correct size.
+        // 6. Apply metadata (use modDataObjAvu/Meta).
+        // 7. Compute checksum on file at rest and compare to the existing one (verification).
+        // 
+        // This API plugin needs to do the same things as rsDataObjClose(...).
+        // We will revisit redirects at a later time.
+
+        try {
+            std::string_view json_string(static_cast<const char*>(_input->buf), _input->len);
+
+            log::api::trace("Parsing string into JSON ... [string={}]", json_string);
+
+            const auto json_input = json::parse(json_string);
+            const auto l1desc_index = json_input.at("file_descriptor").get<int>();
+
+            log::api::trace("Checking file descriptor ...");
+
+            if (l1desc_index < 0 || l1desc_index >= NUM_L1_DESC) {
+                log::api::error("Invalid file descriptor [error_code={}, fd={}].", BAD_INPUT_DESC_INDEX, l1desc_index);
+                return BAD_INPUT_DESC_INDEX;
+            }
+
+            const auto& l1desc = L1desc[l1desc_index];
```

Comment:
should this plugin be using the L1 descriptor table at all given the L1 descriptor information was provided by the client?

Reply:
there is also the dataObjInfo from the catalog?